src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -1915,6 +1917,30 @@ LogicalResult ONNXConvOp::inferShapes(
       (hasBias && !B().getType().isa<RankedTensorType>()))
     return success();
 
+  auto kernelShape = kernel_shape();
+  if (!kernelShape.has_value()) {
+    // Deduce shape from weight input.
+    auto xTy = X().getType().cast<RankedTensorType>();
+    auto xShape = xTy.getShape();
+    auto weightTy = W().getType().cast<RankedTensorType>();
+    auto weightShape = weightTy.getShape();
+    // Number of spatial dimensions.
+    auto spatialOffset = 2;
+    int32_t spatialRank = xShape.size() - spatialOffset;
+    SmallVector<int64_t, 2> defaultVals;
+    for (int i = 0; i < spatialRank; ++i)
+      defaultVals.emplace_back(weightShape[spatialOffset + i]);
+    // Convert to ArrayRef, then build attribute, then store attribute.
+    ArrayRef<int64_t> defaultRefs(defaultVals);
+    auto builder = mlir::Builder(getContext());
+    kernel_shapeAttr(builder.getI64ArrayAttr(defaultRefs));
+    kernelShape = kernel_shape();
+  }
+
+  // Process strides, dilations, and pads.
+  LogicalResult res = processConvTypeParams<>(this, X());
```

Comment:
Looking at the comment before ONNXConvOp::inferShapes, it looks like the intension is to not change the operation's attributes during shape inference but the output's shape only. As I see it, it is still debating. Perhaps, you can raise this issue in our next weekly meeting.

Reply:
I think I will take a shot at rewriting this as a canonicalization

Reply:
We were initially rewriting attributes, but found it easier to simple compute it into shape inference. 
The main problem is that when you change attributes, its a "compiler constant only" situation. But there are cases where it is not compile time constant (e.g the padding), so I found it simpler to have the pads reside into the shape inference helper object, so that we can always fetch it from there, as opposed to "if it is constant, get it from the attribute, if its run time, get it from elsewhere".

I am happy to have this as a topic of discussion in our weekly meeting.

Reply:
Looking at the code, it looks like we have changed the attributes in the past, so it make sense to continue to do so. So indeed putting it into canonicalization may make sense.